### ISSUE TITLE
Add a way to check whether we are currently JITing into `jax.extend`

### ISSUE THREAD

**Issue by patrick-kidger** [google/jax#18544]
As discussed: it would be nice to have a way to have a way to check if we are currently jitting. As an "advanced" not-quite-public feature this could go in `jax.extend`.

Important features:
- be more robust than the current `isinstnace(jnp.array(0) + 1, jax.core.Tracer)` hack, which will pick up on any dynamic trace;
- operate outside of primitive rules, e.g. a primitive with a lowering rule that just raises an error would not be enough.

**Comment by mattjj** [comment]
How about a way to check whether we're currently being staged out to a jaxpr? There are reasons we do that other than `jax.jit`, e.g. for some other higher-order primitives like `jax.remat` or `jax.lax.scan`, or just running a `jax.make_jaxpr`.

**Comment by patrick-kidger** [comment]
Agreed, I think the same check should trigger for both JIT'ing and jaxpr'ing. Otherwise we could silently skip the check by jaxpr'ing and then doing `jit(eval_jaxpr(jaxpr))`.
